flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSink.java

```diff
@@ -323,18 +339,62 @@ public int hashCode() {
                         valueSerialization,
                         keyFieldGetters,
                         valueFieldGetters,
-                        hasMetadata,
+                        hasMetadata(),
                         metadataPositions,
                         upsertMode);
 
         return new FlinkKafkaProducer<>(
                 topic,
                 kafkaSerializer,
                 properties,
-                sinkSemantic.getSemantic(),
+                getSemantic(deliveryGuarantee),
                 FlinkKafkaProducer.DEFAULT_KAFKA_PRODUCERS_POOL_SIZE);
     }
 
+    private int[] getMetadataPositions(List<LogicalType> physicalChildren) {
```

Comment:
Do we need to execute it if `hasMetadata = false`? Can we maybe even encode it through returning `null` and get rid of `hasMetadata` parameter?

Reply:
Sounds like a good idea but I would also consider it a follow-up improvement once we have found a good way to replace the serialization schema and only have to support one.